### DISCUSSION TITLE
Creating new  user types in SB

### DISCUSSION THREAD

**Discussion by vinod303** [Sunbird-Lern/Community#117] [Query]
**Context**:
For  a SB tenant,  we need to setup different User Types / Personas (instead of Student, teacher etc..)

1. I have created those new User Types using the form Create APIs
example of new User Types:  Farmer, FLW etc..

2. Following is the output of form read api after the previous step

   `// "result": {
        "form": {
            "type": "config",
            "subtype": "userty
         pe",
            "action": "get",
            "component": "portal",
            "framework": "*",
            "data": {
                "templateName": "userType",
                "action": "get",
                "fields": [
                    {
                        "code": "public",
                        "name": "Public",
                        "label": "frmelmnts.lbl.public",
                        "visibility": true,
                        "image": "guest-img1.svg",
                        "index": 0,
                        "searchFilter": [
                            "Public"
                        ]
                    },
                    {
                        "code": "farmer",
                        "name": "Farmer",
                        "visibility": true,
                        "label": "frmelmnts.lbl.farmer",
                        "image": "guest-img2.svg",
                        "index": 1,
                        "searchFilter": [
                            "Public",
                            "Farmer"
                        ]
                    },
                    {
                        "code": "flw",
                        "name": "Field Level Worker",
                        "index": 3,
                        "visibility": true,
                        "label": "frmelmnts.lbl.flw",
                        "image": "guest-img5.svg",
                        "searchFilter": [
                            "Public",
                            "Farmer",
                            "Field Level Worker"
                        ]
                    },`

3. Now when I login as user, => click on Profile => Edit preferences, selected the Role as "farmer".  
And when try to save, the User Update Api returns the error 
   **"Invalid value farmer for parameter userType. Please provide a valid value."**

Please help me understand, is there any other configuration needs to be done to make that work? Or is there any other places that i need to add the newly created roles?

cc: @reshmi-nair @rhwarrier

**Comment by rhwarrier** [comment]
Looks like it is being validated someplace, and failing. Is the same value reflecting for the forms during user onboarding as well?
@reshmi-nair  can you pls check ?

**Reply by vinod303** [reply]
@rhwarrier  yes, all the newly created roles displays during onboarding.  This problem is only when a user select these roles from the Profile and save.
Thanks

**Reply by Jayaprakash8887** [reply]
@vinod303 Pls restart Learner service and try

**Reply by vinod303** [reply]
@Jayaprakash8887  , that did not work . Facing the same error.

**Reply by Jayaprakash8887** [reply]
@vinod303 pls confirm if the form API connection configuration in learner service is correct.

**Reply by vinod303** [reply]
@Jayaprakash8887  i think configurations are good. Do i need to check anything specific in config ?
Also, i had noticed the same behaviour in other instance(preprod for example) also sometime back.
cc: @raghupathiguduri

**Reply by Jayaprakash8887** [reply]
@vinod303 Kindly help us with logs from learner-service when the exception occurs. Also, kindly mention release-version used just in case required.

**Reply by vinod303** [reply]
hi @Jayaprakash8887 ,
Pls find the following from the learner log. These log indicates that, it is validating against some already configured values. 
But, i am not sure, which api i should use/where should I add these new userTypes. 
My wild guess is "tenantPreference", but not sure.
Can u please guide?


**# 1.**
`learner-6ddb59f66d-59ns7 learner 2023-05-18T11:37:06.417254854+05:30 {"timestamp":"2023-05-18T06:07:06.417Z","msg":"{\"eid\":\"LOG\",\"ets\":1684390026417,\"ver\":\"3.0\",\"mid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"actor\":{\"id\":\"1abf8eb5-9e02-440b-9adb-217efafed5af\",\"type\":\"User\"},\"context\":{\"channel\":\"01379001379364044836\",\"pdata\":{\"id\":\"production.production.portal\",\"pid\":\"learner-service\",\"ver\":\"4.6.0\"},\"env\":\"User\",\"did\":\"f1f7f65990814401e137069a779f011a\",\"cdata\":[{\"id\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"level\":\"info\",\"type\":\"Api_access\",\"message\"**:\"Available User Type for stateCode:KA are [parent, administrator, other, teacher, student]\",\"params\":[]}}"**,"lname":"TelemetryEventLogger","tname":"application-akka.actor.most-used-one-dispatcher-404","level":"INFO","HOSTNAME":"learner-6ddb59f66d-59ns7","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}`

**# 2**
`learner-6ddb59f66d-59ns7 learner 2023-05-18T11:37:06.417444895+05:30 {"timestamp":"2023-05-18T06:07:06.417Z","msg":"{\"eid\":\"ERROR\",\"ets\":1684390026417,\"ver\":\"3.0\",\"mid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"actor\":{\"id\":\"1abf8eb5-9e02-440b-9adb-217efafed5af\",\"type\":\"User\"},\"context\":{\"channel\":\"01379001379364044836\",\"pdata\":{\"id\":\"production.production.portal\",\"pid\":\"learner-service\",\"ver\":\"4.6.0\"},\"env\":\"User\",\"did\":\"f1f7f65990814401e137069a779f011a\",\"cdata\":[{\"id\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\**"INVALID_PARAMETER_VALUE\",\"stacktrace\":\"Error while processing the message : updateUserV3** org.sunbird.exception.ProjectCommonException.throwClientErrorException(ProjectCommonException.java:96)org.sunbird.actor.user.validator.UserRequestValidator.lambda$validateUserType$3(UserRequestValidator.java:805)java.base/java.util.ArrayList.forEach(ArrayList.java:1541)org.sunbird.actor.user.validator.UserRequestValidator.validateUserType(UserRequestValidator.java:801)org.sunbird.actor.user.UserUpdateActor.validateUserTypeAndSubType(UserUpdateActor.java:363)org.sunbird.actor.user.UserUpdateActor.validateUserTypeAndSubType(UserUpdateActor.java:334)org.sunbird.actor.user.UserUpdateActor.updateUser(UserUpdateActor.java:135)org.sunbird.actor.user.UserUpdateActor.onReceive(UserUpdateActor.java:77)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:20)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive(Actor.scala:517)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\"}}","lname":"TelemetryEventLogger","tname":"application-akka.actor.most-used-one-dispatcher-404","level":"INFO","HOSTNAME":"learner-6ddb59f66d-59ns7","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}
learner-6ddb59f66d-59ns7 learner 2023-05-18T11:37:06.418024866+05:30 {"timestamp":"2023-05-18T06:07:06.417Z","msg":"{\"eid\":\"ERROR\",\"ets\":1684390026417,\"ver\":\"3.0\",\"mid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"actor\":{\"id\":\"1abf8eb5-9e02-440b-9adb-217efafed5af\",\"type\":\"User\"},\"context\":{\"channel\":\"01379001379364044836\",\"pdata\":{\"id\":\"production.production.portal\",\"pid\":\"learner-service\",\"ver\":\"4.6.0\"},\"env\":\"User\",\"did\":\"f1f7f65990814401e137069a779f011a\",\"cdata\":[{\"id\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\"400\",\"stacktrace\":\"org.sunbird.exception.ProjectCommonException.throwClientErrorException(ProjectCommonException.java:9\",\"errtype\":\"INVALID_PARAMETER_VALUE\",\"requestid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\"}}","lname":"TelemetryEventLogger","tname":"Thread-4113","level":"INFO","HOSTNAME":"learner-6ddb59f66d-59ns7","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}
11:42
learner-6ddb59f66d-59ns7 learner 2023-05-18T11:37:06.418252748+05:30 {"timestamp":"2023-05-18T06:07:06.418Z","msg":"{\"eid\":\"ERROR\",\"ets\":1684390026417,\"ver\":\"3.0\",\"mid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"actor\":{\"id\":\"1abf8eb5-9e02-440b-9adb-217efafed5af\",\"type\":\"User\"},\"context\":{\"channel\":\"01379001379364044836\",\"pdata\":{\"id\":\"production.production.portal\",\"pid\":\"learner-service\",\"ver\":\"4.6.0\"},\"env\":\"User\",\"did\":\"f1f7f65990814401e137069a779f011a\",\"cdata\":[{\"id\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\"INVALID_PARAMETER_VALUE\",\"stacktrace\":\"Invalid value farmer for parameter userType. Please provide a valid value. org.sunbird.exception.ProjectCommonException.throwClientErrorException(ProjectCommonException.java:96)org.sunbird.actor.user.validator.UserRequestValidator.lambda$validateUserType$3(UserRequestValidator.java:805)java.base/java.util.ArrayList.forEach(ArrayList.java:1541)org.sunbird.actor.user.validator.UserRequestValidator.validateUserType(UserRequestValidator.java:801)org.sunbird.actor.user.UserUpdateActor.validateUserTypeAndSubType(UserUpdateActor.java:363)org.sunbird.actor.user.UserUpdateActor.validateUserTypeAndSubType(UserUpdateActor.java:334)org.sunbird.actor.user.UserUpdateActor.updateUser(UserUpdateActor.java:135)org.sunbird.actor.user.UserUpdateActor.onReceive(UserUpdateActor.java:77)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:20)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive(Actor.scala:517)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"5dd2a1e8-25a5-d6a5-9b7e-f19b3e93c980\"}}","lname":"TelemetryEventLogger","tname":"Thread-4113","level":"INFO","HOSTNAME":"learner-6ddb59f66d-59ns7","application.home":"/home/sunbird/learner/learning-service-1.0-SNAPSHOT"}`

cc: @rhwarrier @reshmi-nair

**Reply by Jayaprakash8887** [reply]
@vinod303 Above log says that user service is not able to get your latest form API configurations with the newly added userTypes from cache.

https://github.com/Sunbird-Lern/sunbird-lms-service/blob/release-5.2.0/service/src/main/java/org/sunbird/actor/user/validator/UserRequestValidator.java

org.sunbird.actor.user.validator.UserRequestValidator.lambda$validateUserType$3(UserRequestValidator.java:805)

**Reply by vinod303** [reply]
@Jayaprakash8887 thanks!
What could be reason for that? Did i miss anything? Kindly suggest
thanks

**Reply by Jayaprakash8887** [reply]
@vinod303 Please check with ED team if there is a way to verify and update cache with forms.

**Reply by vinod303** [reply]
Thanks @Jayaprakash8887 
@rajeevsathish  can you please guide here?

**Reply by vinod303** [reply]
@rajeevsathish can you please help here?

**Reply by rajeevsathish** [reply]
@vinod303 
I can see that the form is updated properly and also the form read is working fine in portal side that means it is not fetching from cache so have to look into the Learner service only
cc: @Jayaprakash8887

**Reply by Jayaprakash8887** [reply]
@vinod303 seems to be an issue in Sunbird. Please raise a bug/issue. Will address in upcoming releases or request you to contribute the fix. cc: @reshmi-nair @AmiableAnil

**Reply by vinod303** [reply]
@Jayaprakash8887  created the issue https://project-sunbird.atlassian.net/browse/LR-603
cc: @rajeevsathish

**Reply by AmiableAnil** [reply]
@vinod303 Please refer the comment here https://project-sunbird.atlassian.net/browse/LR-603?focusedCommentId=111134
Please let us know if this resolved this issue.

**Reply by vinod303** [reply]
@AmiableAnil  nope. I am still having the same issue in 4.6. 
But, noticed that it works in 5.x.
Not sure why this does not work in 4.6.  Appreciate any input from you on this.

**Reply by Jayaprakash8887** [reply]
@vinod303 Probably you can try taking the relevant code from working version to 4.6 in your forked repo.
